pkg/controllers/management/node/controller.go

```diff
@@ -717,3 +744,81 @@ func (m *Lifecycle) updateRawConfigFromCredential(data map[string]interface{}, r
 	}
 	return nil
 }
+
+func (m *Lifecycle) deleteV1Node(node *v3.Node) (runtime.Object, error) {
+	logrus.Debugf("Deleting v1.node for [%v] node", node.Status.NodeName)
+	if nodehelper.IgnoreNode(node.Status.NodeName, node.Status.NodeLabels) {
+		logrus.Debugf("Skipping v1.node removal for [%v] node", node.Status.NodeName)
+		return node, nil
+	}
+
+	if node.Status.NodeName == "" {
+		return node, nil
+	}
+
+	// get user context k8s client to delete downstream cluster node
+	cluster, err := m.clusterLister.Get("", node.Namespace)
+	if err != nil {
+		return node, err
+	}
+
+	userClient, err := m.clusterManager.UserContext(cluster.Name)
+	if err != nil {
+		return node, err
+	}
+
+	err = userClient.K8sClient.CoreV1().Nodes().Delete(
+		context.TODO(), node.Status.NodeName, metav1.DeleteOptions{})
+	if !kerror.IsNotFound(err) {
+		return node, err
+	}
+
+	return node, nil
+}
+
+func (m *Lifecycle) drainNode(node *v3.Node) error {
+	cluster, err := m.clusterLister.Get("", node.Namespace)
+	if err != nil {
+		return err
+	}
+
+	if !nodehelper.DrainBeforeDelete(node, cluster) {
+		return nil
+	}
+
+	logrus.Infof("node [%s] requires draining before delete, attempting to drain", node.Spec.RequestedHostname)
+	kubeConfig, err := m.getKubeConfig(cluster)
+	if err != nil {
+		return fmt.Errorf("node [%s] error getting kubeConfig", node.Spec.RequestedHostname)
+	}
+
+	if node.Spec.NodeDrainInput == nil {
+		logrus.Debugf("node [%s] has no NodeDrainInput, creating one with 60s timeout",
+			node.Spec.RequestedHostname)
+		node.Spec.NodeDrainInput = &v3.NodeDrainInput{
+			Force:           false,
+			DeleteLocalData: false,
+			GracePeriod:     60,
+			Timeout:         60,
+		}
+	}
+
+	ctx, cancel := context.WithTimeout(m.ctx, time.Duration(node.Spec.NodeDrainInput.Timeout)*time.Second)
+	defer cancel()
+
+	_, msg, err := kubectl.Drain(ctx, kubeConfig, node.Status.NodeName, nodehelper.GetDrainFlags(node))
```

Comment:
Are you sure the error should not be checked first?

Reply:
there were cases i've now fixed a bit (e.g. trying to drain an etcd node) that put me into a really bad place. ultimately eating this error puts us right back to where we were when we started this so felt it was best to guarantee we move past the drain and ultimately make the delete call to vsphere. but if the drain never works and you want the node gone i dont think it should block the deleteNode call. If the drain never works and we're catching this error the node will get stuck in a loop and never delete at all. i'm open to retrying 3x times or something? maybe the error bubbles up and you go fix it and the node drains the second time?

Reply:
Yeah but that would require a fairly quick manual intervention, it can probably stay as is in that case.